### ISSUE TITLE
NS1 Provider: Modeling ns1 metadata

### ISSUE THREAD

**Issue by pashap** [hashicorp/terraform#13636]
There has been a struggle to model NS1 record metadata within the ns1 provider. The crux of the problem has to do with the fact that ns1 metadata is used within nested objects of the [record resource](https://github.com/hashicorp/terraform/blob/master/builtin/providers/ns1/resource_record.go#L37) and that its important not to use default values when ns1 metadata fields are absent from the configuration. 

Ideally, the ns1 record resource could contain nested `meta` schemas that represent ns1 metadata:

```tf
provider "ns1" {
}

resource "ns1_zone" "basic" {
  zone = "testzone.test"
}

resource "ns1_record" "basic" {
  zone = "${ns1_zone.basic.zone}"
  domain = "www.${ns1_zone.basic.zone}"
  type = "CNAME"
  ttl = 60
  meta = { 
    up = true
    weight = 10
  }

  answers = {
    answer = "sub1.${ns1_zone.basic.zone}"
    meta = { 
      up = true
      weight = 10
    }
  }
  answers = {
    answer = "sub2.${ns1_zone.basic.zone}"
    meta = { 
      up = false
      weight = 5
    }
  }
  answers = {
    answer = "sub3.${ns1_zone.basic.zone}"
    meta = {  // <- meta.up will be false bc of default value with d.GetOk
      weight = 5
    }
  }

}
```

I have tried "flattening" the ns1 metadata structure, but still run into the same issue of absent fields != default values for `answers`:

```tf
provider "ns1" {
}

resource "ns1_zone" "basic" {
  zone = "testzone.test"
}

resource "ns1_record" "basic" {
  zone = "${ns1_zone.basic.zone}"
  domain = "www.${ns1_zone.basic.zone}"
  type = "CNAME"
  ttl = 60
  // ns1 metadata fields
  up = true
  weight = 10

  answers = {
    answer = "sub1.${ns1_zone.basic.zone}"
    // ns1 metadata fields
    up = true
    weight = 10
  }
  answers = {
    answer = "sub2.${ns1_zone.basic.zone}"
    // ns1 metadata fields 
    up = false
    weight = 5
  }
  answers = {
    answer = "sub3.${ns1_zone.basic.zone}"
    // ns1 metadata fields
    // without setting "up", it is still false as a default for the "answers" schema
    weight = 5
  }

}
```

There is an open [issue](https://github.com/hashicorp/terraform/issues/5694) addressing how to get `d.GetOk` to properly detect absence from config vs default values. 

Are there any suggestions for how to model nested schemas within a resource that need to differentiate between a field being absent and its default?

### References

- https://github.com/hashicorp/terraform/issues/5694
- https://github.com/hashicorp/terraform/pull/10058#discussion_r87617034

**Comment by catsby** [comment]
Hello – 

The issue https://github.com/hashicorp/terraform/issues/5694 is a frustrating one for sure. I'm not sure we have a complete solution for that yet. 

As a workaround, do you think you could define `up` to be a `TypeString` instead of `TypeBool`? It's a bit odd for sure, but Terraform will convert the boolean value of `true` in configurations to `1` and `false` to `0`, and save them as strings as well, so the value is safe in `parseBool`. 

I wrote an example implementation in a "fake" resource here:

- https://gist.github.com/catsby/9a220e2f394a7d6776e66d6191a850c8

Conceptually it's this:

- https://play.golang.org/p/lFYNT4VHD_

In summary, omitting the value will get you `""`, and `true` or `"true"` (string) will get you a boolean value. 

I think that should work here, but please let me know what you think!

**Comment by catsby** [comment]
I'm going to close this for now. The crux of the issue is #5694, hopefully my idea above is a viable workaround. 

Thanks!

**Comment by pashap** [comment]
@catsby Thanks for the detailed explanation/proposed solution! This looks like it will do the trick, as soon as i get time(should be in the next two weeks), i will try implementing it with the NS1 provider. Thanks again for the examples

**Comment by catsby** [comment]
You're very welcome! Thanks for raising the issue, sorry that #5694 is still unresolved 😦
